src/pybind/mgr/mgr_module.py

```diff
@@ -200,6 +204,66 @@ def get_default_choose_args(dump):
         return dump.get('choose_args').get(CRUSHMap.DEFAULT_CHOOSE_ARGS, [])
 
 
+class CLICommand(object):
+    COMMAND_LIST = []
+
+    def __init__(self, prefix, args=None, desc=None, perm=None):
+        self.prefix = prefix
+        self.args = args if args else ""
+        self.args_dict = {}
+        self.desc = desc if desc else ""
+        self.perm = perm if perm else "rw"
+        self.func = None
+        self._parse_args()
+
+    def _parse_args(self):
+        if not self.args:
+            return
+        args = self.args.split(" ")
+        for arg in args:
+            arg_desc = arg.strip().split(",")
+            arg_d = {}
+            for kv in arg_desc:
+                k, v = kv.split("=")
+                if k != "name":
+                    arg_d[k] = v
+                else:
+                    self.args_dict[v] = arg_d
+
+    def __call__(self, func):
+        self.func = func
+        setattr(self.func, '_cli_cmd_', True)
+        self.COMMAND_LIST.append(self)
+        return self.func
+
+    def call(self, mgr, cmd_dict):
+        kwargs = {}
+        for a, d in self.args_dict.items():
+            if 'req' in d and d['req'] == "false" and a not in cmd_dict:
+                continue
+            kwargs[a.replace("-", "_")] = cmd_dict[a]
+        return self.func(mgr, **kwargs)
```

Comment:
Why do you pass `mgr` here, if it's renamed in all cmd functions as `_` and the global `mgr` is used instead?

Reply:
AFAIK only the dashboard module has the mgr module instance accessible by a global variable, and this command decorator is supposed to be used by any mgr module. 

Also, when you are handling a command in the existing implementation, you override the `handle_command` method of the `MgrModule` subclass and you have access to the mgr module instance using the `self` variable. By passing the `mgr` here we are guaranteeing that you have the same context accessible when using the decorator approach.

Reply:
Ok, in this case I'd suggest 2 other possible approaches that don't require a placeholder for mgr instance:
- The singleton `getInstance` approach: `def f(): print id(mgr.getMgr())`.
- Passing the mgr as a function property (functions are objects too). This is however probably the less pythonic one: `def f(): print id(f.mgr)`

If you prefer modifying the function prototype, why not moving the mgr argument to the last position as optional (mgr=None) and allowing for (future-proof) extra attributes using **kwargs? This is closer to the callback-way used in threading module for example:
```python
def f(a, b, c, mgr=None):
  print id(mgr)
```

Reply:
having the mgr param in the first position allows the developer to write the command function as the MgrModule method like in this code:

https://github.com/rjfd/ceph/blob/wip-dashboard-nfs/src/pybind/mgr/ssh-orch/module.py#L353